### PR TITLE
Added new robot model support; Updated Cmake Minimal Version and Cmake Std

### PR DESCRIPTION
**Pull request by drhieu**

### CODE REVIEW

File: roverrobotics_driver/include/roverrobotics_ros_driver.hpp [hunk 1]
```diff
@@ -16,61 +20,63 @@
...
   std::string estop_trigger_topic_;
+  std::string mode_trigger_topic_ = "/mode_toggle";
   std::string estop_reset_topic_;
   std::string robot_status_topic_;
-  float robot_status_frequency;
+  float robot_status_frequency_;
   float robot_status_frequency_max = 60;
-  float robot_status_frequency_min = 5;
-  float robot_odom_frequency;
+  float robot_status_frequency_min_ = 5;
```
Comment (jbrover) [inline roverrobotics_driver/include/roverrobotics_ros_driver.hpp]: why not use 'const' qualifier for these values. Also it is preferable to use ALL CAPS for defined constants.

File: roverrobotics_driver/include/roverrobotics_ros_driver.hpp [hunk 2]
```diff
@@ -16,61 +20,63 @@
...
-  PidGains pidGains_ = {0, 0, 0};
-  float pid_p_max = 1;
-  float pid_p_min = 0;
-  float pid_i_max = 1;
-  float pid_i_min = 0;
-  float pid_d_max = 1;
-  float pid_d_min = 0;
-  bool estop_state = false;
-  bool closed_loop;
+  float pid_p_max_ = 1;
```
Comment (jbrover) [inline roverrobotics_driver/include/roverrobotics_ros_driver.hpp]: same notes; use 'const' qualifier and ALL CAPS for defined constants

File: roverrobotics_driver/launch/mini.launch [hunk 3]
```diff
@@ -0,0 +1,43 @@
+<launch>
+    <arg name="driver_node_name" default="roverrobotics_driver" />
+    <arg name="config_locks" default="$(find roverrobotics_driver)/config/twist_mux_locks.yaml" />
+    <arg name="config_topics" default="$(find roverrobotics_driver)/config/twist_mux_topics_locks.yaml" />
+    <param name="device_port" value="can0" />
+    <param name="comm_type" value="can" />
+    <param name="robot_type" value="mini" />
+    <param name="robot_mode" value="TRACTION_CONTROL" />
```
Comment (jbrover) [inline roverrobotics_driver/launch/mini.launch]: Let's use INDEPENDENT WHEEL as the default mode

File: roverrobotics_driver/launch/pro.launch [hunk 4]
```diff
@@ -3,16 +3,16 @@
...
-    <param name="motor2_id" value="2" />
-    <param name="motor3_id" value="3" />
-    <param name="motor4_id" value="4" />
     <param name="comm_type" value="serial" />
     <param name="robot_type" value="pro" />
     <param name="status_frequency" value="30" />
     <param name="closed_loop_control" value="false" />
     <param name="Kp" value="0.4" />
     <param name="Ki" value="0.7" />
+    <param name="angular_a" value="0" />
```
Comment (jbrover) [inline roverrobotics_driver/launch/pro.launch]: these values look suspicious. Is angular scaling even implemented on the rover pro? IF SO, then the angular_c value should be 1 by default, and a/b should both be 0.

File: roverrobotics_driver/src/roverrobotics_ros_driver.cpp [hunk 5]
```diff
@@ -25,49 +13,57 @@ ROSWrapper::ROSWrapper(ros::NodeHandle *nh) {
     ROS_FATAL("No communication method set, Shutting down Driver Node");
     ros::shutdown();
   }
-  if (!ros::param::get("closed_loop_control", closed_loop)) {
-    ROS_INFO("no 'closed_loop_control' set; using the default value: 'false'");
-    closed_loop = false;
+  std::string mode_name = "OPEN_LOOP";
```
Comment (jbrover) [inline roverrobotics_driver/src/roverrobotics_ros_driver.cpp]: It is my opinion that any default values for variables should be defined in a header file.

File: roverrobotics_driver/src/roverrobotics_ros_driver.cpp [hunk 6]
```diff
@@ -76,18 +72,48 @@ ROSWrapper::ROSWrapper(ros::NodeHandle *nh) {
     ROS_INFO("no 'traction_factor' set; using the default value: '0'");
     odom_traction_factor_ = 0;
   }
+  if (!ros::param::get("angular_a_coef", angular_scaling_params_.a_coef)) {
```
Comment (jbrover) [inline roverrobotics_driver/src/roverrobotics_ros_driver.cpp:77]: these definitely need to be looked at

File: roverrobotics_driver/src/roverrobotics_ros_driver.cpp [hunk 7]
```diff
@@ -76,18 +72,48 @@ ROSWrapper::ROSWrapper(ros::NodeHandle *nh) {
...
+    ROS_INFO("no 'angular_a_coef' set; using the default value: 0");
+    angular_scaling_params_.a_coef = 0;
+  }
+  if (!ros::param::get("angular_b_coef", angular_scaling_params_.b_coef)) {
+    ROS_INFO("no 'angular_b_coef' set; using the default value: 0");
+    angular_scaling_params_.b_coef = 1;
```
Comment (jbrover) [inline roverrobotics_driver/src/roverrobotics_ros_driver.cpp]: inconsistency between print statement and default assignment

File: roverrobotics_driver/src/roverrobotics_ros_driver.cpp [hunk 8]
```diff
@@ -76,18 +72,48 @@ ROSWrapper::ROSWrapper(ros::NodeHandle *nh) {
...
+  }
+  if (!ros::param::get("angular_c_coef", angular_scaling_params_.c_coef)) {
+    ROS_INFO("no 'angular_c_coef' set; using the default value: 1");
+    angular_scaling_params_.c_coef = 0;
```
Comment (jbrover) [inline roverrobotics_driver/src/roverrobotics_ros_driver.cpp]: inconsistency between print statement and assignment

File: roverrobotics_driver/src/roverrobotics_ros_driver.cpp [hunk 9]
```diff
@@ -115,23 +141,23 @@ ROSWrapper::ROSWrapper(ros::NodeHandle *nh) {
     ROS_INFO("no 'status_topic' set; using the default value: '/robot_status'");
     robot_status_topic_ = "/robot_status";
   }
-  if (!ros::param::get("status_frequency", robot_status_frequency)) {
+  if (!ros::param::get("status_frequency", robot_status_frequency_)) {
     ROS_INFO("no 'status_frequency' set; using the default value: '5.00'");
-    robot_status_frequency = 5.00;
+    robot_status_frequency_ = 5.00;
```
Comment (jbrover) [inline roverrobotics_driver/src/roverrobotics_ros_driver.cpp]: same notes, variable defaults should generally not be defined in any .c or .cpp files.

File: roverrobotics_driver/src/roverrobotics_ros_driver.cpp [hunk 10]
```diff
@@ -252,52 +282,88 @@ void ROSWrapper::publishRobotInfo() {
...
+void RobotDriver::callbackModeTrigger(const std_msgs::Bool::ConstPtr &msg) {
+  int mode;
+  sensor_msgs::JoyFeedbackArray a;
+  if (msg->data) {
+    mode = robot_->cycle_robot_mode();
+    ROS_INFO("new mode %d", mode);
+    switch (mode) {
+      case Control::OPEN_LOOP:
+        ROS_INFO("Robot Mode : Open Loop");
+        // a.set_led = true;
```
Comment (jbrover) [inline roverrobotics_driver/src/roverrobotics_ros_driver.cpp:319]: is this prototype code? should it be removed?

File: roverrobotics_input_manager/scripts/ps4_mapper.py [hunk 11]
```diff
@@ -24,31 +24,41 @@ def __init__(self):
...
+        self._pub_squ = rospy.Publisher('/mode_toggle', Bool, queue_size=1)
+        self._pub_triangle = rospy.Publisher(
+            '/joystick/triangle', Bool, queue_size=1, latch=True)
+        self._pub_circle = rospy.Publisher(
+            '/soft_estop/reset', Bool, queue_size=1)  # , latch =True)
+        self._pub_cross = rospy.Publisher(
+            '/soft_estop/trigger', Bool, queue_size=1)  # , latch =True)
+        self._pub_trim = rospy.Publisher(
+            '/trim_increment', Float32, queue_size=1)
         self._trim_incre_value = rospy.get_param('~trim_increment_value', 0.01)
```
Comment (jbrover) [inline roverrobotics_input_manager/scripts/ps4_mapper.py:48]: this type of value should be defined in ALL_CAPS in .py files, preferably at the top of the .py file

File: roverrobotics_input_manager/scripts/ps4_mapper.py [hunk 12]
```diff
@@ -70,43 +80,58 @@ def cb_status(self, msg):
...
-	    rospy.loginfo('Linear Scale is at %f' ,self._scales["linear"]["x"])
-	    self.buttonpressed = True
+                self._feedback.set_rumble = False
+                self._pub_feedback.publish(self._feedback)
+        trim_msg = 0
+        self._pub_trim.publish(trim_msg)
+
+        if self.togglebuttonpressed:  # Debounce mode
+            self.counter2 += 1
+            if self.counter2 == 200:
```
Comment (jbrover) [inline roverrobotics_input_manager/scripts/ps4_mapper.py]: this is a kluge

File: roverrobotics_input_manager/scripts/ps4_mapper.py [hunk 13]
```diff
@@ -70,43 +80,58 @@ def cb_status(self, msg):
...
-                trim_msg = -self._trim_incre_value
+                trim_msg = .01
             elif msg.button_l1:
-                trim_msg = +self._trim_incre_value
+                trim_msg = -.01
             self._pub_trim.publish(trim_msg)
             self.buttonpressed = True
         elif self.buttonpressed:  # Debounce
             self.counter += 1
             if self.counter == 50:
```
Comment (jbrover) [inline roverrobotics_input_manager/scripts/ps4_mapper.py]: magic numbers

File: roverrobotics_input_manager/scripts/ps4_mapper.py [hunk 14]
```diff
@@ -70,43 +80,58 @@ def cb_status(self, msg):
         if (msg.button_l1 or msg.button_r1) and self.buttonpressed is False:
             trim_msg = Float32()
             if msg.button_r1:
-                trim_msg = -self._trim_incre_value
+                trim_msg = .01
```
Comment (jbrover) [inline roverrobotics_input_manager/scripts/ps4_mapper.py]: magic numbers

File: roverrobotics_driver/include/roverrobotics_ros_driver.hpp [hunk 15]
```diff
@@ -16,61 +20,63 @@
...
   ros::Publisher robot_status_publisher_;  // publish robot state (battery,
-                                           // estop_status, speed)
-  ros::Publisher robot_odom_publisher_;    // publish odometry
+  // estop_status, speed)
+  ros::Publisher feedback_publisher_;
+  ros::Publisher robot_odom_publisher_;  // publish odometry
   // parameter variables
   std::string speed_topic_;
   std::string estop_trigger_topic_;
+  std::string mode_trigger_topic_ = "/mode_toggle";
```
Comment (jbrover) [inline roverrobotics_driver/include/roverrobotics_ros_driver.hpp]: should this be 'const'? does it change?

**Review by jbrover — CHANGES_REQUESTED** [review]

File: roverrobotics_driver/launch/mini.launch (hunk 3 above)
Comment (drhieu) [reply roverrobotics_driver/launch/mini.launch]: changed for Pro2 as well

File: roverrobotics_driver/launch/pro.launch (hunk 4 above)
Comment (drhieu) [reply roverrobotics_driver/launch/pro.launch]: Not sure; would have to check librover;

File: roverrobotics_driver/src/roverrobotics_ros_driver.cpp [hunk 16]
```diff
@@ -0,0 +1,383 @@
...
+  if (pidGains_.kp < PID_P_MIN_) {
+    ROS_WARN("pidGains_.Kp is too low, changing to: %f", PID_P_MIN_);
+    pidGains_.kp = PID_P_MIN_;
+  } else if (pidGains_.kp > PID_P_MAX_) {
+    ROS_WARN("pidGains_.Kp is too high, changing to: %f", PID_P_MAX_);
+    pidGains_.kp = PID_P_MAX_;
+  }
+  if (!ros::param::get("Ki", pidGains_.ki)) {
+    pidGains_.ki = PID_I_DEFAULT_;
+    ROS_INFO("no 'Ki' set; using the default value: '.0'");
```
Comment (jbrover) [inline roverrobotics_driver/src/roverrobotics_ros_driver.cpp]: oops

**Review by jbrover — CHANGES_REQUESTED** [review]

**Review by jbrover — APPROVED** [review]

**Pull request merged**